### ISSUE TITLE
Perpendicular velocities in the Sod problem should be zero 

### ISSUE THREAD

**Issue by shibersag** [STEllAR-GROUP/octotiger#366]
## Expected Behavior

When running the Sod problem, velocities (and momentums) perpendicular to the shock should remain exactly zero, i.e. the -y and z velocities when the shock aligned across the x axis (which is the default behavior).

## Actual Behavior

When running on GPUs the perpendicular velocities have some small values.
This happened before when the hydro AM correction was set on, but when the correction was set off, in the past the problem just disappeared.

## Steps to Reproduce the Problem

Run octoitger from master with the following parameters:
problem=sod
odt=0.02
stop_time=0.021
gravity=off
max_level=3
disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
cuda_number_gpus=1
cuda_streams_per_gpu=32
cuda_buffer_capacity=1
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=CUDA
hydro_host_kernel_type=LEGACY

See in the output:
              sx 2.559705e-03 1.038547e-02
              sy 1.222462e-18 5.680512e-18
              sz 1.215541e-18 5.619879e-18

However, when running without CUDA:
problem=sod
odt=0.02
stop_time=0.021
gravity=off
max_level=3
disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
cuda_number_gpus=1
cuda_streams_per_gpu=32
cuda_buffer_capacity=1
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
#monopole_device_kernel_type=CUDA
#multipole_device_kernel_type=CUDA
#hydro_device_kernel_type=CUDA
hydro_host_kernel_type=LEGACY

The output is:
              sx 2.559705e-03 1.038547e-02
              sy 0.000000e+00 0.000000e+00
              sz 0.000000e+00 0.000000e+00
as it should be.

## Specifications

Current version of Octotiger from master

**Comment by G-071** [comment]
Interesting. This does not happen with the sod test that we have in the test problem's folder (though that one has a max_level=1).

As for the actual difference: It seems to be still within machine precision. 

But it is odd: If you run the scenario for more time-steps with` --odt=1.00 --stop_time=1.21` get similar results with the CPU-only version on master as well as the benchmark version of Octo-Tiger.
With similar, I mean that instead of 0, I get the following results with the benchmark version:
```
 sy 0.000000e+00 0.000000e+00
 sz 5.549214e-16 6.677689e-16
 zx 1.387514e-16 1.923567e-16 
```
sy seems to behave as you mentioned it should, but neither sz nor zx does (which should be 0.0 as well, right?). Though, the values are still extremely close to zero.

**Comment by shibersag** [comment]
We found out that specifying the runtime parameter:

cuda_buffer_capacity=1024

Can fix this bug in some cases. 

For other cases you get sy zero as should be but sz still greater than zero.

This can imply that we have some other bug which may or may not be relevant to this particular bug.
A further investigation is need regarding the secondary issue.

**Comment by shibersag** [comment]
I tried PR#376, and the hydro_host_kernel_type=DEVICE_ONLY seems to work well.
I tested cases of a uniform grid as the Sod tests in the benchmarks paper have been done with a uniform grid only.

However, when the simulation is performed on a long test both on CPU+GPU the errors are adding up well above machine-precision. I attach below config files and outputs for the two cases (run on CPU+GPU and run on GPU only using DEVICE_ONLY):

case 1 (DEVICE_ONLY):
---

ini file -

problem=sod
odt=0.02
stop_time=20
#stop_step=200
gravity=off
max_level=4
unigrid=1
#disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
reflect_bc=1
n_species=1
cuda_number_gpus=1
cuda_streams_per_gpu=32
cuda_buffer_capacity=1
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=CUDA
hydro_host_kernel_type=DEVICE_ONLY


final output -

L1, L2
             rho 1.438357e-01 1.524064e-01
            egas 4.338213e-01 4.358711e-01
             tau 4.346091e-01 4.355058e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762242e-01 3.766278e-01
              sy 0.000000e+00 0.000000e+00
              sz 0.000000e+00 0.000000e+00
              zx 0.000000e+00 0.000000e+00
              zy 4.791708e-03 7.478406e-03
              zz 4.791708e-03 7.478406e-03
           spc_1 1.438357e-01 1.524064e-01
Writing analytic.silo
Start timings report...
...:
   Total: 12641
   Computation: 11293.6 (89.341 %)
   Regrid: 748.109 (5.91811 %)
   Computation + Regrid: 12041.7 (95.2591 %)
Finished timings report!
---


case 2 (VC):
---

ini file -

problem=sod
odt=0.02
stop_time=20
#stop_step=200
gravity=off
max_level=4
unigrid=1
#disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
reflect_bc=1
n_species=1
cuda_number_gpus=1
cuda_streams_per_gpu=32
cuda_buffer_capacity=1
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=CUDA
hydro_host_kernel_type=VC


final output -

L1, L2
             rho 1.435012e-01 1.528907e-01
            egas 4.338213e-01 4.358593e-01
             tau 4.346102e-01 4.355016e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762116e-01 3.766138e-01
              sy 4.695255e-05 6.372605e-05
              sz 4.549716e-05 6.164135e-05
              zx 1.853740e-05 2.693617e-05
              zy 4.795290e-03 7.475993e-03
              zz 4.795103e-03 7.477264e-03
           spc_1 1.435012e-01 1.528907e-01
Writing analytic.silo
Start timings report...
...:
   Total: 12954.7
   Computation: 11612.1 (89.6367 %)
   Regrid: 748.846 (5.78051 %)
   Computation + Regrid: 12361 (95.4172 %)
Finished timings report!

---

And of course in a CPU only run (hydro_device_kernel_type=OFF) the final output is as should be -
L1, L2
             rho 1.435051e-01 1.522987e-01
            egas 4.338213e-01 4.358827e-01
             tau 4.346082e-01 4.355100e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762264e-01 3.766304e-01
              sy 0.000000e+00 0.000000e+00
              sz 0.000000e+00 0.000000e+00
              zx 0.000000e+00 0.000000e+00
              zy 4.791162e-03 7.479971e-03
              zz 4.791162e-03 7.479971e-03
           spc_1 1.435051e-01 1.522987e-01
Writing analytic.silo
Start timings report...
...:
   Total: 30651
   Computation: 29276.6 (95.5157 %)
   Regrid: 752.948 (2.45652 %)
   Computation + Regrid: 30029.5 (97.9722 %)
Finished timings report!

**Comment by G-071** [comment]
@shibersag 
Was Octo-Tiger recompiled with OCTOTIGER_WITH_FAST_FP_CONTRACT=OFF for the CPU+GPU run?

**Comment by shibersag** [comment]
I didn't test those FAST_FP flags yet, that would be my next step. I just used the default value which I think is ON

**Comment by shibersag** [comment]
Compiling with OCTOTIGER_WITH_FAST_FP_CONTRACT=OFF also fixes this issue and the running time is not by much longer. I attach config file and output below.

ini file - 
problem=sod
odt=0.02
stop_time=20
#stop_step=200
gravity=off
max_level=4
unigrid=1
#disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
reflect_bc=1
n_species=1
cuda_number_gpus=1
cuda_streams_per_gpu=32
cuda_buffer_capacity=1
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=CUDA
hydro_host_kernel_type=VC

output -
L1, L2
             rho 1.435051e-01 1.522987e-01
            egas 4.338213e-01 4.358827e-01
             tau 4.346082e-01 4.355100e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762264e-01 3.766304e-01
              sy 0.000000e+00 0.000000e+00
              sz 0.000000e+00 0.000000e+00
              zx 0.000000e+00 0.000000e+00
              zy 4.791162e-03 7.479971e-03
              zz 4.791162e-03 7.479971e-03
           spc_1 1.435051e-01 1.522987e-01
Writing analytic.silo
Start timings report...
...:
   Total: 13022
   Computation: 11702.1 (89.8646 %)
   Regrid: 750.862 (5.76612 %)
   Computation + Regrid: 12453 (95.6308 %)
Finished timings report!

**Comment by G-071** [comment]
@shibersag 

Should we enforce that Octo-Tiger was compiled with OCTOTIGER_WITH_FAST_FP_CONTRACT=OFF in case both CPU and GPU kernels are active (i.e. DEVICE_ONLY is not used)? I can add a check for that at the application startup when the options are processed. That would be a simple modification in PR #376

**Comment by shibersag** [comment]
I also made some tests with Kokkos, where we basically see the same behavior. When running only on CPUs or only on GPUs, the velocities are zeros, otherwise they have some non-zero values which seem to increase with time.

It is also interesting to note, though unrelated to this particular issue, that the Kokkos runs have longer running times compare to Cuda runs (and compare to Vc in the CPUs only run). Below are the ini file and the outputs.


case 1 (DEVICE_ONLY)
---

ini file -
problem=sod
odt=0.02
stop_time=20
#stop_step=10
gravity=off
max_level=4
unigrid=1
#disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
reflect_bc=1
n_species=1
cuda_number_gpus=1
cuda_streams_per_gpu=32
#cuda_buffer_capacity=10
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=KOKKOS_CUDA
hydro_host_kernel_type=DEVICE_ONLY

output - 
L1, L2
             rho 1.438357e-01 1.524064e-01
            egas 4.338213e-01 4.358711e-01
             tau 4.346091e-01 4.355058e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762242e-01 3.766278e-01
              sy 0.000000e+00 0.000000e+00
              sz 0.000000e+00 0.000000e+00
              zx 0.000000e+00 0.000000e+00
              zy 4.791708e-03 7.478406e-03
              zz 4.791708e-03 7.478406e-03
           spc_1 1.438357e-01 1.524064e-01
Writing analytic.silo
Start timings report...
...:
   Total: 21775.1
   Computation: 20429.1 (93.8185 %)
   Regrid: 745.651 (3.42433 %)
   Computation + Regrid: 21174.7 (97.2428 %)
Finished timings report!


case 2 (KOKKOS+KOKKOS_CUDA) - 
---

ini file -
problem=sod
odt=0.02
stop_time=20
#stop_step=10
gravity=off
max_level=4
unigrid=1
#disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
reflect_bc=1
n_species=1
cuda_number_gpus=1
cuda_streams_per_gpu=32
#cuda_buffer_capacity=10
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=KOKKOS_CUDA
hydro_host_kernel_type=KOKKOS

output -
L1, L2
             rho 1.435008e-01 1.528955e-01
            egas 4.338213e-01 4.358603e-01
             tau 4.346101e-01 4.355019e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762171e-01 3.766197e-01
              sy 4.256193e-05 5.784712e-05
              sz 4.405317e-05 5.972602e-05
              zx 1.735566e-05 2.535463e-05
              zy 4.793531e-03 7.478027e-03
              zz 4.794012e-03 7.476476e-03
           spc_1 1.435008e-01 1.528955e-01
Writing analytic.silo
Start timings report...
...:
   Total: 19570.1
   Computation: 18216.4 (93.0829 %)
   Regrid: 743.861 (3.801 %)
   Computation + Regrid: 18960.3 (96.8839 %)
Finished timings report!

case 3 (KOKKOS+OFF)
---

ini file -
problem=sod
odt=0.02
stop_time=20
#stop_step=10
gravity=off
max_level=4
unigrid=1
#disable_diagnostics=on
xscale=0.5
correct_am_hydro=0
reflect_bc=1
n_species=1
cuda_number_gpus=1
cuda_streams_per_gpu=32
#cuda_buffer_capacity=10
monopole_host_kernel_type=VC
multipole_host_kernel_type=VC
monopole_device_kernel_type=CUDA
multipole_device_kernel_type=CUDA
hydro_device_kernel_type=OFF
hydro_host_kernel_type=KOKKOS

output -
L1, L2
             rho 1.435051e-01 1.522987e-01
            egas 4.338213e-01 4.358827e-01
             tau 4.346082e-01 4.355100e-01
             pot 0.000000e+00 0.000000e+00
              sx 3.762264e-01 3.766304e-01
              sy 0.000000e+00 0.000000e+00
              sz 0.000000e+00 0.000000e+00
              zx 0.000000e+00 0.000000e+00
              zy 4.791162e-03 7.479971e-03
              zz 4.791162e-03 7.479971e-03
           spc_1 1.435051e-01 1.522987e-01
Writing analytic.silo
Start timings report...
...:
   Total: 34717.7
   Computation: 33389.3 (96.1736 %)
   Regrid: 744.164 (2.14347 %)
   Computation + Regrid: 34133.5 (98.3171 %)
Finished timings report!

**Comment by shibersag** [comment]
> @shibersag
> 
> Should we enforce that Octo-Tiger was compiled with OCTOTIGER_WITH_FAST_FP_CONTRACT=OFF in case both CPU and GPU kernels are active (i.e. DEVICE_ONLY is not used)? I can add a check for that at the application startup when the options are processed. That would be a simple modification in PR #376

@G-071,
Sure, sounds good, Let's add this check

**Comment by G-071** [comment]
Done!

**Comment by diehlpk** [comment]
@dmarce1 and @shibersag could we try to finish this pull request by Firday?

**Comment by shibersag** [comment]
@G-071, should I do anymore tests before closing this issue?

**Comment by G-071** [comment]
@shibersag Only if you noticed anything odd during you previous tests that calls for more tests :-)  
    
From my end, the smaller tests I ran (and that are now run each commit via Jenkins) look alright at least!

**Comment by shibersag** [comment]
Resolved by PR #376
